WalletWasabi/Wallets/TransactionFeeProvider.cs

```diff
@@ -0,0 +1,117 @@
+using NBitcoin;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.Blockchain.TransactionProcessing;
+using WalletWasabi.Extensions;
+using WalletWasabi.Logging;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class TransactionFeeProvider : PeriodicRunner
+{
+	private readonly TimeSpan _maximumDelay = TimeSpan.FromMinutes(2);
+
+	public TransactionFeeProvider(WasabiHttpClientFactory httpClientFactory) : base(TimeSpan.FromSeconds(10))
+	{
+		HttpClient = httpClientFactory.NewHttpClient(httpClientFactory.BackendUriGetter, Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+	}
+
+	public event EventHandler? RequestedFeeArrived;
+
+	public ConcurrentDictionary<uint256, Money> FeeCache { get; } = new();
+	public ConcurrentQueue<uint256> Queue { get; } = new();
+
+	private IHttpClient HttpClient { get; }
+
+	private async Task FetchTransactionFeeAsync(uint256 txid, CancellationToken cancellationToken)
+	{
+		using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(20));
+		using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+		try
+		{
+			var response = await HttpClient.SendAsync(HttpMethod.Get, $"api/v{Helpers.Constants.BackendMajorVersion}/btc/Blockchain/get-transaction-fee?transactionId={txid}", null, linkedCts.Token).ConfigureAwait(false);
+
+			if (response.StatusCode != HttpStatusCode.OK)
+			{
+				await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+			}
+
+			Money fee = await response.Content.ReadAsJsonAsync<Money>().ConfigureAwait(false);
+
+			if (!FeeCache.TryAdd(txid, fee))
+			{
+				throw new InvalidOperationException($"Failed to cache {txid} with fee: {fee}");
+			}
+
+			RequestedFeeArrived?.Invoke(this, EventArgs.Empty);
```

Comment:
What if instead of `EventArgs.Empty` you send in `txid` and `FeeRate`, so at the receiveing side you can just call like `UpdateFees()` instead of updating the whole transaction, this way I think you are free from a Func parameter and TransactionSummary reference.

Reply:
Yesss, thank you. I changed the approach we talked about yesterday, but the core is the same. [d0a1abf](https://github.com/zkSNACKs/WalletWasabi/pull/11679/commits/d0a1abfb3713ec3972bc72c41d690c71a4811c4f)